src/Servers/Kestrel/Core/src/Internal/Http/PathNormalizer.cs

```diff
@@ -50,193 +52,144 @@ public static string DecodePath(Span<byte> path, bool pathEncoded, string rawTar
-                    {
-                        // B.  if the input buffer begins with a prefix of "/./" or "/.",
-                        //     where "." is a complete path segment, then replace that
-                        //     prefix with "/" in the input buffer; otherwise,
-                        src += 2;
-                        continue;
-                    }
-
-                    break;
-                default:
-                    ch2 = *(src + 1);
-                    ch3 = *(src + 2);
-                    ch4 = *(src + 3);
+                return src.Length;
+            }
+            if (nextDotSegmentIndex < 0)
+            {
+                // Copy the remianing src to dst, and return.
+                src[readPointer..].CopyTo(src[writtenLength..]);
+                writtenLength += src.Length - readPointer;
+                return writtenLength;
+            }
+            else if (nextDotSegmentIndex > 0)
+            {
+                // Copy until the next segment excluding the trailer.
+                Unsafe.CopyBlock(ref src[writtenLength], ref src[readPointer], (uint)nextDotSegmentIndex);
+                writtenLength += nextDotSegmentIndex;
+                readPointer += nextDotSegmentIndex;
+            }
 
-                    if (ch2 == ByteDot && ch3 == ByteDot && ch4 == ByteSlash)
+            // Case of /../ or /./
+            var remainingLength = src.Length - readPointer;
+            if (remainingLength > 3)
+            {
+                var nextIndex = readPointer + 2;
+                if (src[nextIndex] == ByteSlash)
+                {
+                    readPointer = nextIndex;
+                }
+                else if (MemoryMarshal.CreateSpan(ref src[nextIndex], 2).StartsWith(dotSlash))
+                {
+                    // Remove the last segment and replace the path with /
+                    var lastIndex = MemoryMarshal.CreateSpan(ref src[0], writtenLength).LastIndexOf(ByteSlash);
+
+                    // Move write pointer to the end of the previous segment without / or to start position
+                    writtenLength = int.Max(0, lastIndex);
+
+                    // Move the read pointer to the next segments beginning including /
+                    readPointer += 3;
+                }
+                else
+                {
+                    // Not a dot segment, copy the matched /. and bump the read pointer
+                    slashDot.CopyTo(src[writtenLength..]);
+                    writtenLength += 2;
+                    readPointer = nextIndex;
+                }
+            }
+            // Ending with /.. or /./
+            else if (remainingLength == 3)
+            {
+                var nextIndex = readPointer + 2;
+                if (src[nextIndex] == ByteSlash)
+                {
+                    // Replace the /./ segment with a closing /
+                    src[writtenLength++] = ByteSlash;
```

Comment:
The write isn't needed here? Since `src[writtenLength..] == "/./"` we just need to increment `writtenLength`.

Reply:
I think it is `src[readPointer..] == "/./"`, we are not sure what is at `writtenLength`, because a previous `/../` segment like: `/a/.././` could cause an offset